### PR TITLE
Feature - 39263 Playwright New POM Support

### PR DESCRIPTION
**Pull request by IamRanjeetSingh**
Thank you for your contribution.
Before submitting this PR, please make sure:

- [ ] PR description and commit message should describe the changes done in this PR
- [ ] Verify the PR is pointing to correct branch i.e. Release or Beta branch if the code fix is for specific release , else point it to master
- [ ] Latest Code from master or specific release branch is merged to your branch
- [ ] No unwanted\commented\junk code is included
- [ ] No new warning upon build solution
- [ ] Code Summary\Comments are added to my code which explains what my code is doing
- [ ] Existing unit test cases are passed
- [ ] New Unit tests are added for your development
- [ ] Sanity Tests are successfully executed for New and Existing Functionality
- [ ] Verify that changes are compatible with all relevant browsers and platforms.
- [ ] After creating pull request there should not be any conflicts
- [ ] Resolve all Codacy comments
- [ ] Builds and checks are passed before PR is sent for review
- [ ] Resolve code review comments
- [ ] Update the Help Library document to match any feature changes

## Summary by CodeRabbit

- **New Features**
	- Enhanced browser interaction by adding support for handling windows, tabs, frames, and script injections.
	- Introduced methods for generating XPath from HTML elements and creating HTML elements from browser elements.
	- New capabilities for JavaScript execution and element retrieval through `IBrowserTab`.
	- Added extensive support for learning and extracting web page elements for Page Object Model (POM) automation in `POMLearner`.
	- Improved element locating by name, tag name, and relative XPath in `PlaywrightBrowserTab`.

- **Improvements**
	- Refined error handling and interaction logic for browser elements in `PlaywrightDriver`.
	- Enhanced visual testing support with new methods for visual element interaction.

### CODE REVIEW

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/POM/POMLearner.cs [hunk 1]
```diff
@@ -0,0 +1,737 @@
...
+            _htmlDocument = htmlDocument;
+            _browserElementProvider = browserElementProvider;
+            _pomSetting = pomSetting;
+            _xpathImpl = xpathImpl;
+        }
+
+        private static XPathHelper NewXPathHelper(IXPath xpathImpl)
+        {
+            List<string> importantProperties = ["POMLearner", "Web"];
+            return new XPathHelper(xpathImpl, importantProperties);
```
[the remark this answers is not in the document]
Comment (IamRanjeetSingh) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/POM/POMLearner.cs:55]: this is valid syntax

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/POM/POMLearner.cs [hunk 2]
```diff
@@ -0,0 +1,737 @@
...
+                {
+                    return null;
+                }
+                return pomSetting.ElementLocatorsSettingsList.FirstOrDefault(l => l.LocateBy == locateBy);
+            }
+
+            List<ElementLocator> locators = new(new WebPlatform().GetLearningLocators());
+            foreach (ElementLocator locator in locators)
+            {
+                switch (locator.LocateBy)
```
[the remark this answers is not in the document]
Comment (IamRanjeetSingh) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/POM/POMLearner.cs:302]: we don't want to do that

File: Ginger/GingerCoreCommon/UIElement/AppWindow.cs [hunk 3]
```diff
@@ -25,7 +25,7 @@ public enum eWindowType
             Windows,
             PowerBuilder,
             ASCFForm,
-            SeleniumWebPage,
+            WebPage,
```
[the remark this answers is not in the document]
Comment (IamRanjeetSingh) [reply Ginger/GingerCoreCommon/UIElement/AppWindow.cs:28]: those are commented code

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Playwright/PlaywrightDriver.cs [hunk 4]
```diff
@@ -426,64 +440,994 @@
...
+            ThrowIfClosed();
+            return _browser;
+        }
+
+        public List<AppWindow> GetAppWindows()
+        {
+            //TODO: unhighlight last highlighted element
+            IEnumerable<AppWindow> appWindows = Task.Run(() => GetAppWindowsAsync().Result).Result;
+            return new List<AppWindow>(appWindows);
+        }
```
[the remark this answers is not in the document]
Comment (IamRanjeetSingh) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Playwright/PlaywrightDriver.cs:465]: we have handled deadlock scenario

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/GingerWebDriver.cs [hunk 5]
```diff
@@ -51,5 +83,410 @@
...
+            try
+            {
+                object? output = await tab.ExecuteJavascriptAsync("return GingerLibLiveSpy.IsLiveSpyExist();");
+                string? outputString = output?.ToString();
+                if (!string.IsNullOrEmpty(outputString))
+                {
+                    isInjected = outputString;
+                }
+            }
+            catch (Exception) { }
```
[the remark this answers is not in the document]
Comment (IamRanjeetSingh) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/GingerWebDriver.cs:284]: handled

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/GingerWebDriver.cs [hunk 6]
```diff
@@ -51,5 +83,410 @@
...
+
+            //don't know why we execute below 2 scripts, copying it from SeleniumDriver
+            await tab.ExecuteJavascriptAsync("define_GingerLibLiveSpy();");
+            await tab.ExecuteJavascriptAsync($"return GingerLibLiveSpy.AddScript({JavaScriptHandler.GetJavaScriptFileContent(JavaScriptHandler.eJavaScriptFile.jquery_min)});");
+        }
+
+        private protected Task InjectScriptAsync(IBrowserTab tab, string script)
+        {
+            string injectionableScript = PrepareScriptForInjection(script);
+            //return tab.ExecuteJavascriptAsync(injectionableScript);
```
[the remark this answers is not in the document]
Comment (IamRanjeetSingh) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/GingerWebDriver.cs:307]: handled

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Playwright/PlaywrightDriver.cs [hunk 7]
```diff
@@ -426,64 +440,994 @@
...
+                elemInfo.Y += parentElementLocation.Y;
+
+                return elemInfo;
+            }
+            catch (Exception ex)
+            {
+                Reporter.ToLog(eLogLevel.ERROR, "Failed to Get Element At Point", ex);
+                return null;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (IamRanjeetSingh) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Playwright/PlaywrightDriver.cs:1450]: we handled deadlock scenario

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/POM/POMLearner.cs [hunk 8]
```diff
@@ -0,0 +1,800 @@
...
+                Width = size.Width,
+                Height = size.Height,
+                X = position.X,
+                Y = position.Y,
+                HTMLElementObject = htmlNode,
+                XPath = htmlNode.XPath,
+                IsAutoLearned = true,
+                ScreenShotImage = await GetElementScreenshotAsync(browserElement),
+                Locators = [],
+                Properties = [
```
[the remark this answers is not in the document]
Comment (IamRanjeetSingh) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/POM/POMLearner.cs:172]: this is valid syntax

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/POM/POMLearner.cs [hunk 9]
```diff
@@ -0,0 +1,800 @@
...
+                string[] xpathSegments = htmlElementInfo.Path.Split('/');
+                lastXPathSegment = xpathSegments[xpathSegments.Length - 1];
+            }
+            string xpath = string.Empty;
+            if (!lastXPathSegment.Contains("frame"))
+            {
+                xpath = htmlElementInfo.Path;
+            }
+
+            Stack<HtmlNode> nodes = [];
```
[the remark this answers is not in the document]
Comment (IamRanjeetSingh) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/POM/POMLearner.cs:679]: this is valid syntax

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/POM/POMLearner.cs [hunk 10]
```diff
@@ -0,0 +1,800 @@
...
+            }
+
+            return properties;
+        }
+
+        private IEnumerable<ElementLocator> GenerateRelativeXPathLocators(HTMLElementInfo htmlElementInfo)
+        {
+            if (htmlElementInfo.ElementTypeEnum == eElementType.Svg)
+            {
+                return [];
```
[the remark this answers is not in the document]
Comment (IamRanjeetSingh) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/POM/POMLearner.cs:551]: this is valid syntax

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/POM/POMLearner.cs [hunk 11]
```diff
@@ -0,0 +1,800 @@
...
+            }
+
+            List<ElementLocator> locators = [];
```
[the remark this answers is not in the document]
Comment (IamRanjeetSingh) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/POM/POMLearner.cs:554]: this is valid syntax

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/GingerWebDriver.cs [hunk 12]
```diff
@@ -51,5 +83,410 @@
...
+            await tab.ExecuteJavascriptAsync($"arg => GingerLibLiveSpy.AddScript(arg);", JavaScriptHandler.GetJavaScriptFileContent(JavaScriptHandler.eJavaScriptFile.jquery_min));
+            await tab.ExecuteJavascriptAsync("GingerLibLiveSpy.StartEventListner();");
+        }
+
+        private protected Task InjectScriptAsync(IBrowserTab tab, string script)
+        {
+            string injectionableScript = PrepareScriptForInjection(script);
+            //return tab.ExecuteJavascriptAsync(injectionableScript);
+            return tab.InjectJavascriptAsync(injectionableScript);
+        }
```
[the remark this answers is not in the document]
Comment (IamRanjeetSingh) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/GingerWebDriver.cs:309]: handled

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/GingerWebDriver.cs [hunk 13]
```diff
@@ -51,5 +83,410 @@
...
+                object? output = await tab.ExecuteJavascriptAsync("GingerLibLiveSpy.IsLiveSpyExist();");
+                string? outputString = output?.ToString();
+                if (!string.IsNullOrEmpty(outputString))
+                {
+                    isInjected = outputString;
+                }
+            }
+            catch (Exception) { }
+
+            return !string.Equals(isInjected, "no");
```
[the remark this answers is not in the document]
Comment (IamRanjeetSingh) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/GingerWebDriver.cs:286]: handled

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/GingerWebDriver.cs [hunk 14]
```diff
@@ -51,5 +83,410 @@
...
+                    appWindows.Add(new()
+                    {
+                        Title = title,
+                        WindowType = AppWindow.eWindowType.WebPage,
+                    });
+                }
+            }
+
+            return appWindows;
+        }
```
[the remark this answers is not in the document]
Comment (IamRanjeetSingh) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/GingerWebDriver.cs:111]: unnecessary

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Playwright/PlaywrightBrowserElement.cs [hunk 15]
```diff
@@ -110,9 +371,100 @@
...
+            if (!double.TryParse(coordinates[0], out double x))
+            {
+                throw new Exception($"Unable to get element position, invalid x coordinate ");
+            }
+            if (!double.TryParse(coordinates[1], out double y))
+            {
+                throw new Exception($"Unable to get element position, invalid y coordinate {coordinates[1]}");
+            }
+
+            return new Point((int)x, (int)y);
```
[the remark this answers is not in the document]
Comment (IamRanjeetSingh) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Playwright/PlaywrightBrowserElement.cs:410]: fine for now

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Playwright/PlaywrightBrowserElement.cs [hunk 16]
```diff
@@ -110,9 +371,100 @@
...
```
[the remark this answers is not in the document]
Comment (IamRanjeetSingh) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Playwright/PlaywrightBrowserElement.cs:437]: fine for now

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/POM/POMLearner.cs [hunk 17]
```diff
@@ -0,0 +1,800 @@
...
+        private IEnumerable<ElementLocator> GenerateXPathLocatorsFromUserTemplates(IEnumerable<HtmlAttribute> htmlAttributes)
+        {
+            if (_pomSetting == null || 
+                _pomSetting.relativeXpathTemplateList == null ||
+                _pomSetting.relativeXpathTemplateList.Count <= 0)
+            {
+                return [];
+            }
+
+            List<ElementLocator> locators = [];
```
[the remark this answers is not in the document]
Comment (IamRanjeetSingh) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/POM/POMLearner.cs:617]: this is valid syntax

**Review by Maheshkale447 — APPROVED** [review]

**Pull request merged**